eval.go

```diff
@@ -692,77 +361,289 @@ func attrInfoForStructField(field reflect.StructField) (attrInfo, error) {
+		case uint:
+			result = strconv.AppendUint(nil, uint64(val), 10)
+		case int8:
+			result = strconv.AppendInt(nil, int64(val), 10)
+		case uint8:
+			result = strconv.AppendUint(nil, uint64(val), 10)
+		case int16:
+			result = strconv.AppendInt(nil, int64(val), 10)
+		case uint16:
+			result = strconv.AppendUint(nil, uint64(val), 10)
+		case int32:
+			result = strconv.AppendInt(nil, int64(val), 10)
+		case uint32:
+			result = strconv.AppendUint(nil, uint64(val), 10)
+		case int64:
+			result = strconv.AppendInt(nil, val, 10)
+		case uint64:
+			result = strconv.AppendUint(nil, val, 10)
+		case uintptr:
+			result = strconv.AppendUint(nil, uint64(val), 10)
+		}
+	}
+	if len(result) == 0 {
+		return nil, &userAttrMissingError{attr: attr}
+	}
+	return result, nil
+}
+
+func (t *userTypeInfo) getSemver(v reflect.Value, attr string) (*semver.Version, error) {
+	info, ok := t.fields[attr]
+	if ok && info.asSemver != nil {
+		ver, err := info.asSemver(v)
+		if err != nil {
+			return nil, &userAttrError{attr: attr, err: err}
+		}
+		return ver, nil
+	} else if t.getAttribute != nil {
+		if res, ok := t.getAttribute(v, attr).([]byte); ok {
+			ver, err := parseSemver(string(res))
+			if err != nil {
+				return nil, &userAttrError{attr: attr, err: err}
+			}
+			return ver, nil
+		}
+		if res, ok := t.getAttribute(v, attr).(string); ok {
+			ver, err := parseSemver(res)
+			if err != nil {
+				return nil, &userAttrError{attr: attr, err: err}
+			}
+			return ver, nil
+		}
 	}
-	if tinfo.getAttribute == nil {
-		return attrInfo{
-			asString: func(v reflect.Value) string {
-				return ""
-			},
-			asBytes: func(v reflect.Value) []byte {
-				return nil
-			},
+	return nil, &userAttrMissingError{attr: attr}
+}
+
+func (t *userTypeInfo) getFloat(v reflect.Value, attr string) (float64, error) {
+	info, ok := t.fields[attr]
+	if ok && info.asFloat != nil {
+		res, err := info.asFloat(v)
+		if err != nil {
+			return 0, &userAttrError{attr: attr, err: err}
+		}
+		return res, nil
+	} else if t.getAttribute != nil {
+		val := t.getAttribute(v, attr)
+		switch val := val.(type) {
+		case float64:
+			return val, nil
+		case float32:
+			return float64(val), nil
+		case string:
+			res, err := parseFloat(val)
+			if err != nil {
+				return 0, &userAttrError{attr: attr, err: err}
+			}
+			return res, nil
+		case []byte:
+			res, err := parseFloat(string(val))
+			if err != nil {
+				return 0, &userAttrError{attr: attr, err: err}
+			}
+			return res, nil
+		case int:
+			return float64(val), nil
+		case uint:
+			return float64(val), nil
+		case int8:
+			return float64(val), nil
+		case uint8:
+			return float64(val), nil
+		case int16:
+			return float64(val), nil
+		case uint16:
+			return float64(val), nil
+		case int32:
+			return float64(val), nil
+		case uint32:
+			return float64(val), nil
+		case int64:
+			return float64(val), nil
+		case uint64:
+			return float64(val), nil
+		case uintptr:
+			return float64(val), nil
+		case time.Time:
```

Comment:
This one only describes the float representation of `time.Time` (regardless of what comparator is used), and datetime comparators are comparing floats no matter what the source is. I'd rather not put any restriction logic here, plus I don't see a problem in using a datetime with number comparators.

Reply:
> I don't see a problem in using a datetime with number comparators.

I'm afraid, I do. To mention no more, this is against the spec, which is pretty clear about what types of values are to be accepted in the case of the specific comparators.

Because of the "magic" (auto-conversion) which is going on under the hood, this feature might cause some confusion in itself. So, I think, we especially need consistency here to keep the confusion as low as possible.

Reply:
I don't think it's against the spec. In my opinion, it conforms the spec and by the nature of how it was implemented here, extends the defined possibilities. 

I don't think there's magic under the hood, the `time.Time` type is converted to float by design (which will be documented), therefore it can be used in operations that use float values. Just as we do with `string` now.

I'm not against the spec or anything, I just don't want to introduce unnecessary restrictions.

Reply:
Resolved in private discussion: It's now restricted to only datetime comparators